network/router.go

```diff
@@ -0,0 +1,394 @@
+package network
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/dedis/cothority/log"
+)
+
+// Router handles all networking operations such as:
+// * it listens to incoming connections using a host.Listener methods
+// * open up new connections using  host.Connect's method
+// * dispatch incoming message using a Dispatcher
+// * dispatch outgoing message maintaining a translation
+//   between ServerIdentity <-> address
+// * manage the reconnections of non-working Conn,
+// Most caller should use the creation function like NewTCPRouter(...),
+// NewLocalRouter(...) then use the Host such as:
+// `router.Start() // will listen for incoming Conn and block`
+// `router.Stop() // will stop the listening and the managing of all Conn`
+type Router struct {
+	// id is our own ServerIdentity
+	id *ServerIdentity
+	// address is the real-actual address used by the listener.
+	address Address
+	// Dispatcher is used to dispatch incoming message to the right recipient
+	Dispatcher
+
+	host Host
+	// connections keeps track of all active connections with the translation
+	// use of an array because it happens that some connections are opened at
+	// the same time on both endpoints, and thus registered one after the other,
+	// erasing the first conn.
+	connections map[ServerIdentityID][]Conn
+	connsMut    sync.Mutex
+
+	// boolean flag indicating that the router is already clos{ing,ed}
+	isClosed  bool
+	closedMut sync.Mutex
+
+	// we wait that all handleConn routines are done
+	wg    sync.WaitGroup
+	wgMut sync.Mutex
+}
+
+// NewRouter returns a fresh Router giving its identity, and the host we want to
+// use.
+func NewRouter(own *ServerIdentity, h Host) *Router {
+	r := &Router{
+		id:          own,
+		connections: make(map[ServerIdentityID][]Conn),
+		host:        h,
+		Dispatcher:  NewBlockingDispatcher(),
+	}
+	r.address = h.Address()
+	r.host.Listening()
+	h.Listening()
+	return r
+}
+
+// Start will start the listening routine of the underlying Host. It is a
+// blocking call until the listening is done (by calling r.Stop()).
+func (r *Router) Start() {
+	// The function given to the listener  does the exchange of ServerIdentity
+	// and pass the connection along to the router.
+	err := r.host.Listen(func(c Conn) {
+		dst, err := r.exchangeServerIdentity(c)
+		if err != nil {
+			log.Error("ExchangeServerIdentity failed:", err)
+			if err := c.Close(); err != nil {
+				log.Error("Couldn't close secure connection:",
+					err)
+			}
+			return
+		}
+		// pass it along
+		r.launchHandleRoutine(dst, c)
+	})
+	if err != nil {
+		log.Error("Error listening:", err)
+	}
+}
+
+// Stop will stop the listening routine, and stop any routine of handling
+// connections. Calling r.Start(), then r.Stop() then r.Start() again leads to
+// an undefined behaviour. Callers should most of the time re-create a fresh
+// Router.
+func (r *Router) Stop() error {
+	var err error
+	if r.host.Listening() {
+		err = r.host.Stop()
+	}
+	err2 := r.stopHandling()
```

Comment:
I propose to include `stopHandling`, `reset` and `close` in here, as they are only used once.

I'd also remove the `closedMut` and use `connsMut` instead.